x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/version/EsqlVersion.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.version;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.VersionId;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public enum EsqlVersion implements VersionId<EsqlVersion> {
+    /**
+     * Breaking changes go here until the next version is released.
+     */
+    NIGHTLY(Integer.MAX_VALUE, Integer.MAX_VALUE, "😴"),
+    PARTY_POPPER(2024, 4, "🎉");
+
+    static final Map<String, EsqlVersion> VERSION_MAP_WITH_AND_WITHOUT_EMOJI = versionMapWithAndWithoutEmoji();
+
+    private static Map<String, EsqlVersion> versionMapWithAndWithoutEmoji() {
+        Map<String, EsqlVersion> stringToVersion = new LinkedHashMap<>(EsqlVersion.values().length * 2);
+
+        for (EsqlVersion version : EsqlVersion.values()) {
+            EsqlVersion existingVersionForKey = null;
+            existingVersionForKey = stringToVersion.put(version.versionStringWithoutEmoji(), version);
+            assert existingVersionForKey == null;
+            existingVersionForKey = stringToVersion.put(version.toString(), version);
+            assert existingVersionForKey == null;
+        }
```

Comment:
```suggestion
        for (EsqlVersion version : EsqlVersion.values()) {
            assert null == stringToVersion.put(version.versionStringWithoutEmoji(), version);
            assert null == stringToVersion.put(version.toString(), version);
        }
```

Reply:
I think I actually prefer an explicit throw here rather than just an assert.  This should run once per instance of Elasticsearch, I don't think we need the performance gain of being able to disable asserts in production.

Reply:
I'm not entirely sure how `assert` works, but I'd be worried that putting these onto one line will make it not perform the `put` if assertions are disabled. FWIW, it doesn't really matter if it runs it or not - it'll slow down a reader either way. But the `assert` on it's own line is fin.

You could make it a hard test. The extra test on startup is probably not going to hurt us.

Reply:
This suggestion looks cleaner, but is not equivalent; disabling assertions breaks the version parsing altogether as no version strings are added to the `stringToVersion` map anymore.

Reply:
I'll add an explicit `throw new AssertionError(...)`.